ocaml/tests/test_repository_helpers.ml

```diff
@@ -3608,6 +3608,448 @@ module GetLatestUpdatesFromRedundancy = Generic.MakeStateless (struct
+        , ([RestartToolstack], [], [RestartDeviceModel])
+        )
+      ; ( ([RestartDeviceModel], [RestartDeviceModel])
+        , ([RestartDeviceModel], [], [])
+        )
+      ; ( ([RestartDeviceModel], [RestartDeviceModel; RestartToolstack])
+        , ([RestartToolstack; RestartDeviceModel], [], [])
+        )
+        (* group: unapplied:[RestartDeviceModel; RestartToolstack], 20 ~ 24 *)
+      ; ( ([RestartDeviceModel; RestartToolstack], [])
+        , ([], [], [RestartToolstack; RestartDeviceModel])
+        )
+      ; ( ([RestartDeviceModel; RestartToolstack], [RebootHost])
+        , ([RebootHost], [RestartToolstack; RestartDeviceModel], [])
+        )
+      ; ( ([RestartDeviceModel; RestartToolstack], [RestartToolstack])
+        , ([RestartToolstack], [], [RestartDeviceModel])
+        )
+      ; ( ([RestartDeviceModel; RestartToolstack], [RestartDeviceModel])
+        , ([RestartDeviceModel], [], [RestartToolstack])
+        )
+      ; ( ( [RestartDeviceModel; RestartToolstack]
+          , [RestartDeviceModel; RestartToolstack]
+          )
+        , ([RestartToolstack; RestartDeviceModel], [], [])
+        )
+      ]
+end)
+
+module MergeLivepatchFailures = Generic.MakeStateless (struct
+  module Io = struct
+    (* (previous_failures, succeeded, failed) *)
+    type input_t =
+      Guidance.t list * (Livepatch.component list * Livepatch.component list)
+
+    (* failure_guidances *)
+    type output_t = Guidance.t list
+
+    let string_of_input_t (previous_failures, (succeeded, failed)) =
+      Fmt.(
+        str "%a" Dump.(pair (list string) (pair (list string) (list string)))
+      )
+        ( List.map Guidance.to_string previous_failures
+        , ( List.map Livepatch.string_of_component succeeded
+          , List.map Livepatch.string_of_component failed
+          )
+        )
+
+    let string_of_output_t failure_guidances =
+      Fmt.(str "%a" Dump.(list string))
+        (List.map Guidance.to_string failure_guidances)
+  end
+
+  let transform (previous_failures, (succeeded, failed)) =
+    merge_livepatch_failures ~previous_failures ~succeeded ~failed
+
+  let tests =
+    let open Guidance in
+    let open Livepatch in
+    `QuickAndAutoDocumented
+      [
+        (* 0 ~ 4 *)
+        (([], ([], [])), [])
+      ; (([], ([], [Xen])), [RebootHostOnXenLivePatchFailure])
+      ; (([], ([], [Kernel])), [RebootHostOnKernelLivePatchFailure])
+      ; ( ([], ([], [Xen; Kernel]))
+        , [RebootHostOnXenLivePatchFailure; RebootHostOnKernelLivePatchFailure]
+        )
+      ; (([], ([Xen], [])), []) (* 5 ~ 9 *)
+      ; (([], ([Xen], [Kernel])), [RebootHostOnKernelLivePatchFailure])
+      ; (([], ([Kernel], [])), [])
+      ; (([], ([Kernel], [Xen])), [RebootHostOnXenLivePatchFailure])
+      ; (([], ([Xen; Kernel], [])), [])
+      ; ( ([RebootHostOnXenLivePatchFailure], ([], []))
+        , [RebootHostOnXenLivePatchFailure]
+        )
+        (* 10 ~ 14 *)
```

Comment:
These kind of tests would probably be easier to write and maintain using property based testing.
Although the manually written ones are useful too (e.g. a property test could check whether the manual ones match expectations or not).

Reply:
How about adding a new test case and in the case, the number of exhausted inputs is to be verified against the number of sub cases in this test?
It has to be done manually to map the auto-generated input to an expected output.